### ISSUE TITLE
Support setting namespaces in RunInference

### ISSUE THREAD

**Issue by yeandy** [apache/beam#21721]
### What needs to happen?

Right now, the name for namespaces in RunInference is hardcoded to `RunInference` for base.py and `sklearn_inference.py` and `RunInferencePytorch` for `pytorch_inference.py` in the `get_metrics_namespace` function. However, users may want more control over how the namespaces. For example, one could have two classification models that use the same model loader, and want different namespaces to distinguish between the two. This would also help with metrics tracking in our community dashboard. With the old solution, there isn't a way to do this. If we can allow passing in a parameter like `metrics_namespace` into the RunInference transform, then users would be able to set their own namespace. 

example:
RunInference(model_loader, metrics_namespace=”my_custom_namespace”)

### Issue Priority

Priority: 2

### Issue Component

Component: sdk-py-core

Subtask of issue https://github.com/apache/beam/issues/21435

**Comment by yeandy** [comment]
@ryanthompson591 @AnandInguva

**Comment by AnandInguva** [comment]
Can we make `get_metrics_namespace` static?

**Comment by yeandy** [comment]
you mean with a `@staticmethod` decorator? or static as in hard coded?

**Comment by AnandInguva** [comment]
Yep. `@staticmethod` decorator. Because in the benchmark tests, If I need to filter the metrics using namespace, I can just get the `namespace` by calling `PytorchRunInference.get_name_space()`

**Comment by AnandInguva** [comment]
.take-issue

**Comment by ryanthompson591** [comment]
Just an idea --- How about making the default_namespace static? That way you can get it, but if the user specifies something separate then they can.

**Comment by AnandInguva** [comment]
@yeandy @ryanthompson591 

Do we want the namespace to be defined on `ModelHandler` or `RunInference` transform?  

Question :
- On `ModelHandler`, user can define a namespace while creating the ModelHandler object and can pass this ModelHandler object to the RunInference transform.
- What happens when there are two `RunInference` transforms that use the same `ModelHandler` and the two transforms need different namespaces?

To solve this, we can move the methods related to namespaces into the `RunInference` base class.

**Comment by AnandInguva** [comment]
cc @damccorm

**Comment by ryanthompson591** [comment]
It makes the most sense to me to put the namespace on the base model handler.  This puts the user in full control.

Some outstanding questions would be what would happen with wrapper model handlers - like KeyedModelHandler.'

I think the obvious thing would be to make some sort of default behavior that wrappers should do.  Something like passing the namespace overrides/prefixes to the wrapped model handlers.

I'm not sure if the base model handler should prefix name spaces or not.  Something like:
handler = MyModelHandler(name_space_override='my_defined_namespace')
-> namespace is set to my_defined_namespace
or
handler = SklearnModelHandler(name_space_prefix='some_prefix_')
-> namespace is set to some_prefix_sklearn_namespace

Anand, I think there are a few options how to do this. What do you think about outlining them in a short doc, and then we can pick one?

**Comment by damccorm** [comment]
I think my personal lean would be:

1) Leave the metrics namespace definition in the `ModelHandler` classes. These have the advantage of being model specific; for example, the `PytorchModelHandlerTensor` has a metrics namespace `RunInferencePytorch`. Getting that level of model-specificity is harder to do elsewhere.
2) Add a method (or probably just a class parameter) to override the namespace on the `RunInference` base class. This gives users full control. I'd actually also lean towards making this a postfix to the default metrics namespace instead of a full override since this makes usage tracking easier for runners that can do that (probably just dataflow at this point). It also guarantees that similar metrics are grouped together alphabetically. It does remove some user control though and I could be convinced this is a bad idea.

> Anand, I think there are a few options how to do this. What do you think about outlining them in a short doc, and then we can pick one?

Agreed - so far we've collectively proposed 3 different options :) it would be good to call out pros/cons side by side.

**Comment by AnandInguva** [comment]
Sure. I will create a doc. I agree with Danny and actually worked up on PR[1] based on that. 

[1] https://github.com/apache/beam/pull/23182
